.tito/lib/git_annex_spec_builder.py

```diff
@@ -0,0 +1,168 @@
+
+        # Set to path to srpm once we build one.
+        self.srpm_location = None
+
+    def _get_build_version(self):
+        self.spec_version = ''
+        self.spec_release = ''
+        build_version = get_spec_version_and_release(
+            self.start_dir,
+            find_spec_like_file(self.start_dir)
+        )
+        self.build_tag = build_version
+        return build_version
+
+    def _setup_sources(self):
+        """
+        Create a copy of the git source for the project at the point in time
+        our build tag was created.
+
+        Created in the temporary rpmbuild SOURCES directory.
+        """
+        self._create_build_dirs()
+
+        debug('SETUP SOURCES')
+        working_path = os.getcwd() if self.relative_project_dir in os.path.join(os.getcwd(), '') else os.path.join(os.getcwd(), self.relative_project_dir)
+        debug("working_path: %s" % working_path)
+
+        for directory, unused, filenames in os.walk(working_path):
+            debug('WALK')
+            dir_artifacts_with_path = [os.path.join(directory, f) for f in filenames]
+
+            debug(dir_artifacts_with_path)
+            for artifact in dir_artifacts_with_path:
+                debug("  Copying source file %s" % artifact)
+                if os.path.isfile(artifact):
+                    shutil.copy(artifact, self.rpmbuild_gitcopy)
+                    shutil.copy(artifact, self.rpmbuild_sourcedir)
+
+        # NOTE: The spec file we actually use is the one exported by git
+        # archive into the temp build directory. This is done so we can
+        # modify the version/release on the fly when building test rpms
+        # that use a git SHA1 for their version.
+        self.spec_file_name = os.path.basename(find_spec_like_file(self.rpmbuild_sourcedir))
+        self.spec_file = os.path.join(
+            self.rpmbuild_sourcedir, self.spec_file_name)
+
+        self.old_cwd = os.getcwd()
+        if self.relative_project_dir not in os.path.join(os.getcwd(), ''):
+            os.chdir(os.path.join(self.old_cwd, self.relative_project_dir))
+
+        # NOTE: 'which' may not be installed... (docker containers)
+        (status, output) = getstatusoutput("which git-annex")
+        if status != 0:
+            msg = "Please run '%s' as root." % self.package_manager.install(["git-annex"])
+            error_out('%s' % msg)
+
+        run_command("git-annex lock")
+        annexed_files = run_command("git-annex find --include='*'").splitlines()
+        run_command("git-annex get")
+        run_command("git-annex unlock")
+        debug("  Annex files: %s" % annexed_files)
+
+        for annex in annexed_files:
+            debug("Copying unlocked file %s" % annex)
+            if os.path.isfile(os.path.join(self.rpmbuild_gitcopy, annex)):
+                os.remove(os.path.join(self.rpmbuild_gitcopy, annex))
+            shutil.copy(annex, self.rpmbuild_gitcopy)
+
+        self._lock()
+        os.chdir(self.old_cwd)
```

Comment:
I'm wondering if a chdir context manager is overkill:
```python
from contextlib import contextmanager

@contextmanager
def chdir(path):
    old = os.getcwd()
    os.chdir(path)
    try:
        yield
    finally:
        os.chdir(old)

with chdir('/tmp'):
    assert os.getcwd() == '/tmp'
assert os.getcwd() != '/tmp'
```

Reply:
I can try it. As most of this, code examples and style were lifted from tito.